pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -182,22 +167,16 @@ async fn connection_manager_loop_step(
                 }
             },
 
-            // Got a new update from etcd
-            broker_update = broker_subscription.value_updates.recv() => {
+            // Got a new update from the broker
+            broker_update = broker_subscription.message() => {
                 match broker_update {
-                    Some(broker_update) => walreceiver_state.register_timeline_update(broker_update),
-                    None => {
-                        info!("Broker sender end was dropped, ending current broker loop step");
-                        // Ensure to cancel and wait for the broker subscription task end, to log its result.
-                        // Broker sender end is in the broker subscription task and its drop means abnormal task completion.
-                        // First, ensure that the task is stopped (abort can be done without errors on already stopped tasks and repeated multiple times).
-                        broker_subscription.watcher_handle.abort();
-                        // Then, wait for the task to finish and print its result. If the task was finished before abort (which we assume in this abnormal case),
-                        // a proper error message will be printed, otherwise an abortion message is printed which is ok, since we're signalled to finish anyway.
-                        cleanup_broker_connection(
-                            (&mut broker_subscription.watcher_handle).await,
-                            walreceiver_state,
-                        );
+                    Ok(Some(broker_update)) => walreceiver_state.register_timeline_update(broker_update),
+                    Err(e) => {
+                        error!("broker subscription failed: {e}");
+                        return ControlFlow::Continue(());
+                    }
+                    Ok(None) => {
+                        error!("broker subscription stream ended"); // can't happen
```

Comment:
Why "can't happen"? gRPC request with streaming can break at any time, for example if there are issues with network connection or the server decided to shut down the request. We should consider this and write code for retries, the etcd version was doing it fine.

Also, it seems that we don't have tests with broker unavailability/restarts, such a test would be good for testing this retry logic.

Reply:
> Why "can't happen"?

We never end the stream. In case of connection failure, it would be
```
2022-11-28T06:45:12.889282Z ERROR wal_connection_manager{tenant=de2dfd66cce92d8c4e5266e5e42725e7 timeline=692a8f4f575b0ce8bc4801d601b8abaf}: broker subscription failed: status: Unknown, message: "error reading a body from connection: broken pipe", details: [], metadata: MetadataMap { headers: {} }
```
neighbour arm. But we still do retries regardless whether it was failure or end of stream.


Yes, adding tests would be nice. I'd probably do that after adding to sk endpoint listing current walsender connections, it would make test more interesting (terminate broker and sk which is currently connected, ensure things still work).

Reply:
Hmm, it seems that I misunderstood the scope of the `return ControlFlow::Continue(());` and got triggered by it. Now it looks ok, but I'll try to restart broker in the docker-compose setup to see what will happen.